snapshotter/demux/cache/cache.go

```diff
@@ -15,29 +15,166 @@ package cache
 
-// Cache defines the interface for a snapshotter caching mechanism.
-type Cache interface {
-	// Retrieves the snapshotter from the underlying cache using the provided
-	// fetch function if the snapshotter is not currently cached.
-	Get(ctx context.Context, key string, fetch SnapshotterProvider) (*proxy.RemoteSnapshotter, error)
+// RemoteSnapshotterCache implements a cache for remote snapshotters.
+type RemoteSnapshotterCache struct {
+	mutex        *sync.RWMutex
+	snapshotters map[string]*proxy.RemoteSnapshotter
 
-	// WalkAll applies the provided function across all cached snapshotters.
-	WalkAll(ctx context.Context, fn snapshots.WalkFunc, filters ...string) error
+	fetch SnapshotterProvider
 
-	// Closes the snapshotter and removes it from the cache.
-	Evict(key string) error
+	oneReaper      *sync.Once
+	evict          chan string
+	evictionPolicy EvictionPolicy
 
-	// Releases the cache's internal resources and closes any cached snapshotters.
-	Close() error
+	stop chan struct{}
+}
+
+// SnapshotterCacheOption is a functional option that operates on a remote snapshotter cache.
+type SnapshotterCacheOption func(*RemoteSnapshotterCache)
+
+// NewRemoteSnapshotterCache creates a new instance with an empty cache and applies any provided caching options.
+func NewRemoteSnapshotterCache(fetch SnapshotterProvider, opts ...SnapshotterCacheOption) *RemoteSnapshotterCache {
+	c := &RemoteSnapshotterCache{
+		mutex:        &sync.RWMutex{},
+		snapshotters: make(map[string]*proxy.RemoteSnapshotter),
+		fetch:        fetch,
+		oneReaper:    &sync.Once{},
+		stop:         make(chan struct{}),
+	}
+
+	for _, opt := range opts {
+		opt(c)
+	}
+
+	return c
+}
+
+// EvictOnConnectionFailure is a caching option for evicting entries from the cache after a failed connection attempt.
+func EvictOnConnectionFailure(dial func(context.Context, string) (net.Conn, error), frequency time.Duration) func(*RemoteSnapshotterCache) {
+	return func(c *RemoteSnapshotterCache) {
+		c.evict = make(chan string)
+		c.evictionPolicy = NewEvictOnConnectionFailurePolicy(c.evict, c.stop, dial, frequency)
+		c.launchReaper()
+	}
+}
+
+func (c *RemoteSnapshotterCache) launchReaper() {
+	reaper := func() {
+		go func() {
+			for {
+				s, ok := <-c.evict
+				if !ok {
+					break
+				}
+				c.Evict(s)
+			}
+		}()
+	}
+
+	if c.evictionPolicy != nil {
+		c.oneReaper.Do(reaper)
+	}
+}
+
+// List returns the keys of a snapshotter cache.
+func (c *RemoteSnapshotterCache) List() []string {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+
+	keys := make([]string, 0, len(c.snapshotters))
+	for k := range c.snapshotters {
+		keys = append(keys, k)
+	}
+
+	return keys
+}
+
+// Get fetches and caches the snapshotter for a given key.
+func (c *RemoteSnapshotterCache) Get(ctx context.Context, key string) (*proxy.RemoteSnapshotter, error) {
+	c.mutex.RLock()
+	s, ok := c.snapshotters[key]
+	c.mutex.RUnlock()
+
+	if !ok {
+		c.mutex.Lock()
+		defer c.mutex.Unlock()
+
+		s, ok = c.snapshotters[key]
```

Comment:
We did `s, ok := c.snapshotters[key]` a few lines above. Does it need to do that again?

Reply:
This is our double check lock to prevent a race condition where two requests populate a snapshotter in the cache potentially leaking resources.